design/authorization-handler.md

```diff
@@ -0,0 +1,601 @@
+
+* Expose variables so users can fully customize the text they show for Dynamic Scopes and RAR.
+
+* Allow requesting data from users when a certain type of authorization data is used. For example: when defining a dynamic scope to enforce certain context in tokens, or to define variables in RAR requests.
+
+* Allow users to select which scopes they want to consent. Right now it’s all or none.
+
+The default behaviour of the consent screen will allow multiple fields to be selected / input in the consent screen itself. Some handlers may let users configure this and detach it from the main consent screen so input selection can be done in its own different screen.
+
+### Exposing variables for text customization
+
+We should also strive for giving more flexibility to define the text that is shown in the consent screen, based on the provided data from the authorization requests.
+
+The server could use either [Json Pointer](https://datatracker.ietf.org/doc/html/rfc6901) or [Json Path](https://jsonpath.com/) to access the authorization_details data in a way that users can obtain certain information to use in the consent text.
+
+Example:
+
+An authorization_details object sent to the server.
+```json
+{
+   "type": "payment_initiation",
+   "locations": [
+       "https://example.com/payments"
+   ],
+   "instructedAmount": {
+       "currency": "EUR",
+       "amount": "125.00"
+   }
+}
+```
+
+If the user wanted to show, say, “Allow the client to send 120.00 EUR?”, the actual string that the user would configure in Keycloak would look like:
+
+"Allow the client to send `${instructedAmount.amount} ${instructedAmount.currency}`?” 
+
+### Composable consent screen
+
+
+The consent screen will now be composed of multiple HTML templates that can be configured by each authorization handler.
+
+The base template will be defined by the static scopes that are sent and configured in the server, just like it’s being done right now.
+
+ClientAuthorizationHandlers will be able to configure some predefined templates to add to this.
+
+For now, the following default templates could be considered:
+
+* Dropdown selection consent template: These will allow handlers to request a value from a dropdown. These can contain fixed values or values calculated from the system, such as tenants or groups. For example: When sending the group:* dynamic scope, it would contain all the groups for the requesting user.
+
+* Value selection consent template: These will allow handlers to request a value from the users in order to add it to the consent screen. For example: Transfer amount, source/target account etc.
+
+These selections will be subjected to the policies defined by the handler, which will execute every time there’s a new value being added.
+
+**Question**: Should we also consider further consent screen enhancement that may be out of scope from this proposal now that we are improving it? I’m thinking of showing a client image, or name and UX improvements.
```

Comment:
FYI. Something like this is supported by OIDC specification and we recently have community contribution, which added some of those enhancements to the consent screen. See for more details: https://github.com/keycloak/keycloak/commit/63c9845cb92d558a66f41ead68baff063fb6a03f#diff-3033beba91b6de9a2ae714e742dff4894b2fd9f9caeaea72dffdff0b17e50062

Reply:
I'll add it as an "answer" to the question in the document to keep track of it.